scalalib/test/src/mill/scalalib/HelloWorldTests.scala

```diff
@@ -1341,5 +1395,196 @@ object HelloWorldTests extends TestSuite {
+            "MultiModuleClasspaths/ModMod/bar/unmanaged",
+            "multiModuleClasspaths/modMod/ModMod/bar/compile.dest/classes",
+            //
+            "MultiModuleClasspaths/ModMod/foo/compile-resources",
+            "MultiModuleClasspaths/ModMod/foo/unmanaged",
+            "multiModuleClasspaths/modMod/ModMod/foo/compile.dest/classes",
+            //
+            "MultiModuleClasspaths/ModMod/qux/compile-resources",
+            "MultiModuleClasspaths/ModMod/qux/unmanaged"
+            // We do not include `qux/compile.dest/classes` here, because this is the input
+            // that is required to compile `qux` in the first place
+          )
+        )
+
+      }
+      "modCompile" - workspaceTest(MultiModuleClasspaths) { eval =>
+        // Mostly the same as `modMod` above, but with the dependency
+        // from `qux` to `bar` being a `compileModuleDeps`
+        check(
+          eval,
+          MultiModuleClasspaths.ModCompile.qux,
+          expectedRunClasspath = List(
+            // `utest` is a `runIvyDeps` and not picked up transitively
+            "com/lihaoyi/utest_2.13/0.7.0/utest_2.13-0.7.0.jar",
+            // Because `sourcecode` comes from `ivyDeps`, and the dependency from
+            // `qux` to `bar` is a `compileModuleDeps`, we do not include its
+            // dependencies for `qux`'s `runClasspath`
+            "com/lihaoyi/sourcecode_2.13/0.2.0/sourcecode_2.13-0.2.0.jar",
+            //
+            "org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar",
+            "org/scala-sbt/test-interface/1.0/test-interface-1.0.jar",
+            "org/portable-scala/portable-scala-reflect_2.13/0.1.0/portable-scala-reflect_2.13-0.1.0.jar",
+            "org/scala-lang/scala-reflect/2.13.12/scala-reflect-2.13.12.jar",
+            //
+            "MultiModuleClasspaths/ModCompile/bar/compile-resources",
+            "MultiModuleClasspaths/ModCompile/bar/unmanaged",
+            "MultiModuleClasspaths/ModCompile/bar/resources",
+            "multiModuleClasspaths/modCompile/ModCompile/bar/compile.dest/classes",
+            //
+            "MultiModuleClasspaths/ModCompile/foo/compile-resources",
+            "MultiModuleClasspaths/ModCompile/foo/unmanaged",
+            "MultiModuleClasspaths/ModCompile/foo/resources",
+            "multiModuleClasspaths/modCompile/ModCompile/foo/compile.dest/classes",
+            //
+            "MultiModuleClasspaths/ModCompile/qux/compile-resources",
+            "MultiModuleClasspaths/ModCompile/qux/unmanaged",
+            "MultiModuleClasspaths/ModCompile/qux/resources",
+            "multiModuleClasspaths/modCompile/ModCompile/qux/compile.dest/classes"
+          ),
+          expectedCompileClasspath = List(
+            "com/lihaoyi/geny_2.13/0.4.0/geny_2.13-0.4.0.jar",
+            // We aggregate upstream module `compileIvyDeps` during compilation
+            "com/lihaoyi/sourcecode_2.13/0.2.2/sourcecode_2.13-0.2.2.jar",
```

Comment:
I'm not sure if this is correct: should `compileIvyDeps` be transitive at all? But it is the current behavior

Reply:
`compileIvyDeps` should not be transitive. They only are for local compilation. Downstream consumers need to provide them themselves if needed.

Reply:
It turns out that I was mistaken, and this was a normal `ivyDeps` but in a `compileModuleDeps`.

@lefou it seems you left a comment in here stating that this is necessary. Do you remember the reasoning?

https://github.com/com-lihaoyi/mill/pull/955/files#diff-85ad27993fcc36e23f993baae04e41cb1bfa900502fa0622197d964abca08ddcR103

I'm finding that without propagating `ivyDeps` from the `compileModuleDeps`, a whole bunch of Mill's build stops compiling because e.g. 

1. `A` ends up compiling against module `B` with ivy dependency `C` on it's classpath
2. But if `B` has a method that returns a type from `C`, then `A` will cause the compiler to crash since it needs to have `C` on the classpath to look up and compile against

It seems to me that this should apply equally for `ivyDeps` and `compileIvyDeps`. Currently, a module's `compileClasspath` only picks up `ivyDeps` from it's `compileModuleDeps`, but does not pick up it's `compileIvyDeps`. This has likely not been a problem so far only because `compileIvyDeps` is so uncommonly used.

Given that `ivyDeps` and `compileIvyDeps` have the same problem, should they be inherited by `compileClasspath` in the same way?

Reply:
Ha, I asked myself the same question, when I tried to fix this issue. Unfortunately I don't remember exactly. The timing for me to deep dive into this issue is a bit unlucky. I hope I can later review and think about it. (I also pushed my WIP from last night, just to see the CI results.)

I think we definitely want the handle `compileIvyDeps` and `compileModuleDeps` equally, as the latter is converted to the former once published, and Mill behave the same, disregarding the compile-dependency comes from the local build or from a coursier dependency.

Reply:
sure, this isn't super urgent, so I'll leave it open until you have a chance to think about it

Reply:
I think feel like the best thing to do here is to have all compile dependencies and run dependencies be transitive, but limited to their respective classpaths.

Otherwise, there'll be no end to problems where ModuleA compiling against an upstream ModuleB fails due to ModuleB compiling against ModuleC and exposing it in a type signature. Similarly, for run dependencies, it makes sense that an application that needs e.g. log4j at runtime will continue to need log4j at runtime if used as a library for another application.

It's possible to work around these issues in user land, but it would basically always come down to cherry picking arbitrary dependencies from your upstream modules until things work, or manually traversing the module graph and collecting everything. The former is fragile and the latter is something that it feels like we should automate.

Reply:
There are various out-side givens to incorporate. E.g. we consume dependencies via coursier, so we can't change how transitivity is handled by it, hence, we are forced to adapt our inter-module dependencies handling likewise.

Also, Maven-like `provided` dependencies are useful for compile-time-only APIs, like macros (like `acyclic`?), which should not leak transitively. Or to ensure you to against some older baseline APIs. You don't want to force those old versions transitively, but you want to force your local compilation to use that old version to ensure, you don't accidentally use newer stuff.

Reply:
Good point. We can't change the maven pom scopes, and those have no way to express "transitive compile time dependency". That means that no matter what we do with `moduleDeps`, there's no way to have a transitive compile time dependency published to maven central

I guess in that case, it sounds like we follow the current status quo, where we pick up the transitive `*Deps` of `compile*Deps`, but not the transitive `compile*Deps`.